rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,241 @@
+
+The new `local-overlay` store also uses OverlayFS, but just for the store directory.
+The database is still a regular fresh empty one to start, and instead Nix explicitly knows about the lower store, so it can get any information for the DB it needs from it manually.
+This avoids all 3 downsides:
+
+- Store objects are never deduplicated.
+  OverlayFS ensures that one copy in either store dir layer is enough, and we are careful never to wastefully include a store object in both layers.
+
+- No excess synchronization.
+  local changes are just that: local, not shared with any other consumer.
+  The lower store is never written to (no modifications or even filesystem locks) so any slow NFS write and sync paths should not be encountered.
+
+- No rigidity of the lower store.
+  Since Nix sees both the `overlay-store`'s DB and the lower store, it is possible for it to be aware of and utilize objects that were added to the lower store after the upper store was created.
+
+This gives us a "best of all three worlds" solution.
+
+## Marketing motivation
+
+It is quite common for organizations using Nix to first adopt it behind the scenes.
+That is to say, Nix is used to prepare some artifacts which are then presented to a consumer that need not be aware they were made with Nix.
+Later though, because of Nix's gaining popularity, there may be a desire to reveal its usage so consumers can use Nix themselves.
+Rather than Nix being a controversial tool worth hiding, it can be a popular tool worth exposing.
+Nix-unware usage can still work, but Nix-aware usage can do additional things.
+
+The `local-overlay` store can serve as a crucial tool to bridge these two modes of using Nix.
+The lower store can be however the artifacts were disseminated in the "hidden Nix" first phase of adoption, perhaps with a small tweak to expose the DB / daemon socket if it wasn't before.
+The `local-overlay` store is new, but purely local, separate for each user that wants to use Nix, and completely not impacting any user that doesn't.
+
+By providing the `local-overlay` store, we are essentially completing a reusable step-by-step guide for Nix users to "Nixify their workplace" in a very conscientious and non-disruptive manner.
+
+# Detailed design
+[design]: #detailed-design
+
+## Basic principle
+
+`local-overlay` is a store representing the extension of a lower store with a collection of additional store objects.
+(We don't refer to the upper layer as an "upper store" because it is not self-contained
+--- it doesn't abide by the closure property because objects in the upper layer can refer to objects that are only in the lower layer.)
+
+## Class hierarchy, configuration settings, and initialization
+
+**TODO put in layering diagram, e.g. from blog**
+
+`LocalOverlayStore` is a subclass of `LocalStore` implementing the `local-overlay` store.
+It has additional configuration items for:
+
+ - The lower store, which must be a `LocalFSStore`
```

Comment:
Does this preclude using special lower stores? (ie. the experimental FUSE-backed one)

Reply:
The FUSE-backed one should be okay, as long as it derives from `LocalFSStore`. This requirement just forces it to have a ["real store directory"](https://github.com/NixLayeredStore/nix/blob/ff12cf3b94e2303cb738f34712c0b8660f1261b6/src/libstore/local-fs-store.hh#L32-L41) that we can use as the `lowerdir` of the OverlayFS mount.